src/compose/images.coffee

```diff
@@ -0,0 +1,195 @@
+Promise = require 'bluebird'
+_ = require 'lodash'
+EventEmitter = require 'events'
+logTypes = require '../lib/log-types'
+constants = require '../lib/constants'
+validation = require '../lib/validation'
+
+ImageNotFoundError = (err) ->
+	return "#{err.statusCode}" is '404'
+
+# image = {
+# 	name: image registry/repo:tag
+# 	appId
+# 	serviceId
+# 	serviceName
+# 	imageId (from resin API)
+# 	releaseId
+# 	dependent
+# 	status Downloading, Downloaded, Deleting
+# 	download_progress
+# }
+
+module.exports = class Images extends EventEmitter
+	constructor: ({ @docker, @logger, @db }) ->
+		@imageCleanupFailures = {}
+		# A store of volatile state for images (e.g. download progress), indexed by imageId
+		@volatileState = {}
+
+	reportChange: (imageId, status) ->
+		if status?
+			@volatileState[imageId] ?= { imageId }
+			_.merge(@volatileState[imageId], status)
+			@emit('change')
+		else if imageId? and @volatileState[imageId]?
+			delete @volatileState[imageId]
+			@emit('change')
+
+	fetch: (image, opts) =>
+		onProgress = (progress) =>
+			@reportChange(image.imageId, { download_progress: progress.percentage })
+
+		@normalise(image.name)
+		.then (imageName) =>
+			image = _.clone(image)
+			image.name = imageName
+			@markAsSupervised(image)
+			.then =>
+				@inspectByName(imageName)
+			.catch =>
+				@reportChange(image.imageId, _.merge(_.clone(image), { status: 'Downloading', download_progress: 0 }))
+				Promise.try =>
+					if validation.checkTruthy(opts.delta)
+						@logger.logSystemEvent(logTypes.downloadImageDelta, { image })
+						@docker.rsyncImageWithProgress(imageName, opts, onProgress)
+					else
+						@logger.logSystemEvent(logTypes.downloadImage, { image })
+						@docker.fetchImageWithProgress(imageName, opts, onProgress)
+				.then =>
+					@logger.logSystemEvent(logTypes.downloadImageSuccess, { image })
+					@inspectByName(imageName)
+				.catch (err) =>
+					@logger.logSystemEvent(logTypes.downloadImageError, { image, error: err })
+					throw err
+				.finally =>
+					@reportChange(image.imageId)
+
+	format: (image) ->
+		image.serviceId ?= null
+		image.serviceName ?= null
+		image.imageId ?= null
+		image.releaseId ?= null
+		image.dependent ?= false
+		return _.omit(image, 'id')
+
+	markAsSupervised: (image) =>
+		image = @format(image)
+		@db.upsertModel('image', image, image)
+
+	update: (image) =>
+		image = @format(image)
+		@db.models('image').update(image).where(name: image.name)
+
+	_removeImageIfNotNeeded: (image) =>
+		removed = true
+		@inspectByName(image.name)
+		.catch ImageNotFoundError, (err) ->
+			removed = false
+			return null
+		.then (img) =>
+			if img?
+				@db.models('image').where(name: image.name).select()
+				.then (imagesFromDB) =>
+					if imagesFromDB.length == 1 and _.isEqual(@format(imagesFromDB[0]), @format(image))
+						@docker.getImage(image.name).remove(force: true)
```

Comment:
You can move the `.catch()` here and remove both the `removed` flag and the `if img?` check in the previous block

Reply:
Nice one, done (and added a catchReturn to make it even simpler)